libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt

```diff
@@ -0,0 +1,271 @@
+import com.salesforce.androidsdk.rest.ClientManager.LoginOptions
+import com.salesforce.androidsdk.rest.RestClient.AsyncRequestCallback
+import com.salesforce.androidsdk.rest.RestRequest
+import com.salesforce.androidsdk.rest.RestResponse
+import com.salesforce.androidsdk.security.BiometricAuthenticationManager
+import com.salesforce.androidsdk.security.SalesforceKeyGenerator
+import com.salesforce.androidsdk.ui.OAuthWebviewHelper
+import com.salesforce.androidsdk.util.SalesforceSDKLogger
+import okhttp3.MediaType.Companion.toMediaTypeOrNull
+import okhttp3.RequestBody
+import okhttp3.RequestBody.Companion.toRequestBody
+import org.jetbrains.annotations.VisibleForTesting
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+
+internal class NativeLoginManager(
+    private val clientId: String,
+    private val redirectUri: String,
+    private val loginUrl: String,
+): NativeLoginManager {
+
+    private val accountManager = SalesforceSDKManager.getInstance().userAccountManager
+    private val bioAuthLocked
+        get() = SalesforceSDKManager.getInstance().biometricAuthenticationManager?.locked == true
+    override val shouldShowBackButton: Boolean
+        get() = !bioAuthLocked && accountManager.authenticatedUsers != null
+    override val biometricAuthenticationUsername: String?
+        get() {
+            return if (bioAuthLocked) {
+                accountManager.currentUser.username
+            } else {
+                null
+            }
+        }
+
+    override suspend fun login(username: String, password: String): NativeLoginResult {
+        val trimmedUsername = username.trim()
+        val trimmedPassword = password.trim()
+
+        if (!isValidUsername(trimmedUsername)) {
+            return NativeLoginResult.InvalidUsername
+        }
+
+        if (!isValidPassword(trimmedPassword)) {
+            return NativeLoginResult.InvalidPassword
+        }
+
+        val creds = "$trimmedUsername:$trimmedPassword".toByteArray()
+        val encodedCreds = Base64.encodeToString(creds, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
+        val codeVerifier = SalesforceKeyGenerator.getRandom128ByteKey()
+        val codeChallenge = SalesforceKeyGenerator.getSHA256Hash(codeVerifier)
+
+        val authRequestHeaders = hashMapOf(
+            AUTH_REQUEST_TYPE to NAMED_USER,
+            CONTENT_TYPE to HTTP_POST_CONTENT_TYPE,
+            AUTHORIZATION to AUTHORIZATION_TYPE_BASIC + encodedCreds,
+        )
+        val authRequestBody = createRequestBody(
+            RESPONSE_TYPE to CODE_CREDENTIALS,
+            CLIENT_ID to clientId,
+            REDIRECT_URI to redirectUri,
+            CODE_CHALLENGE to codeChallenge,
+        )
+        val authRequest = RestRequest(
+            RestRequest.RestMethod.POST,
+            RestRequest.RestEndpoint.LOGIN,
+            loginUrl + OAUTH_AUTH_PATH, // Full path for unauthenticated request
```

Comment:
Could this use a Kotlin string template? This should be equivalent, but perhaps it'd be nice just for consistency.

Reply:
🤷 Could be, but I that `loginUrl + OAUTH_AUTH_PATH` is more readable than `"$loginUrl$OAUTH_AUTH_PATH"`.